src/PostgREST/QueryBuilder.hs

```diff
@@ -165,11 +165,25 @@ callProc qi pgArgs returnsScalar selectQuery countQuery countTotal isSingle para
       | paramsAsSingleObject = ("_args_record AS (SELECT NULL)", "$1::json")
       | null pgArgs = (ignoredBody, "")
       | otherwise = (
+          let explicit = pgVer >= pgVersion100
+              rawType a = if explicit
+                then case pgaType a of
+                       "json"  -> "text"
+                       "jsonb" -> "text"
+                       t       -> t
+                else pgaType a
+              conversion = unwords [
+                "SELECT",
+                intercalate ", " ((\a -> pgFmtIdent (pgaName a) <> "::" <> pgaType a) <$> pgArgs),
+                "FROM ("]
+          in
           unwords [
             normalizedBody <> ",",
             "_args_record AS (",
-              "SELECT * FROM json_to_recordset(" <> selectBody <> ") AS _(" <>
-                intercalate ", " ((\a -> pgFmtIdent (pgaName a) <> " " <> pgaType a) <$> pgArgs) <> ")",
+            if explicit then conversion else "",
+            "SELECT * FROM json_to_recordset(" <> selectBody <> ") AS _(" <>
```

Comment:
Try making the change to `jsonb_to_recordset`. Maybe tests pass without the casting to `text` logic above.

Not sure right now if there will be some downsides to using `jsonb_to_recordset`.